// joe-reporter-console/out/lib/console.js
// Generated by CoffeeScript 1.6.2
var ConsoleReporter, cliColor, err, isBrowser, isWindows, _ref,
  __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

isBrowser = typeof window !== "undefined" && window !== null;

isWindows = (typeof process !== "undefined" && process !== null ? (_ref = process.platform) != null ? _ref.indexOf('win') : void 0 : void 0) === 0;

try {
  if (!isBrowser) {
    cliColor = require('cli-color');
  }
} catch (_error) {
  err = _error;
  cliColor = null;
}

ConsoleReporter = (function() {
  ConsoleReporter.prototype.errors = null;

  ConsoleReporter.prototype.config = null;

  function ConsoleReporter(config) {
    var _base, _base1, _base2, _base3, _base4, _base5, _base6, _ref1, _ref2, _ref3, _ref4, _ref5, _ref6, _ref7;

    this.errors || (this.errors = []);
    this.config || (this.config = config || {});
    if ((_ref1 = (_base = this.config).start) == null) {
      _base.start = '';
    }
    if ((_ref2 = (_base1 = this.config).fail) == null) {
      _base1.fail = isWindows ? ' ERR!' : ' ✘  ';
    }
    if ((_ref3 = (_base2 = this.config).pass) == null) {
      _base2.pass = isWindows ? ' OK' : ' ✔  ';
    }
    if ((_ref4 = (_base3 = this.config).sub) == null) {
      _base3.sub = isWindows ? ' > ' : ' ➞  ';
    }
    if ((_ref5 = (_base4 = this.config).failHeading) == null) {
      _base4.failHeading = 'Error #%s:';
    }
    if ((_ref6 = (_base5 = this.config).summaryPass) == null) {
      _base5.summaryPass = "%s/%s tests ran successfully, everything passed";
    }
    if ((_ref7 = (_base6 = this.config).summaryFail) == null) {
      _base6.summaryFail = "FAILURE: %s/%s tests ran successfully; %s failed, %s incomplete, %s errors";
    }
    if (cliColor != null) {
      if (__indexOf.call(typeof process !== "undefined" && process !== null ? process.argv : void 0, '--no-colors') < 0) {
        this.config.fail = cliColor.red(this.config.fail);
        this.config.pass = cliColor.green(this.config.pass);
        this.config.sub = cliColor.black(this.config.sub);
        this.config.failHeading = cliColor.red.underline(this.config.failHeading);
        this.config.summaryPass = cliColor.green.underline(this.config.summaryPass);
        this.config.summaryFail = cliColor.red.bold.underline(this.config.summaryFail);
      }
    }
  }

  ConsoleReporter.prototype.getItemName = function(item) {
    var name;

    name = this.joe.getItemName(item, this.config.sub);
    return name;
  };

  ConsoleReporter.prototype.startSuite = function(suite) {
    var message, name;

    name = this.getItemName(suite);
    if (!name) {
      return this;
    }
    message = "" + name + this.config.start;
    console.log(message);
    return this;
  };

  ConsoleReporter.prototype.finishSuite = function(suite, err) {
    var check, message, name;

    name = this.getItemName(suite);
    if (!name) {
      return this;
    }
    check = (err ? this.config.fail : this.config.pass);
    message = "" + name + check;
    console.log(message);
    return this;
  };

  ConsoleReporter.prototype.startTest = function(test) {
    var message, name;

    name = this.getItemName(test);
    if (!name) {
      return this;
    }
    message = "" + name + this.config.start;
    console.log(message);
    return this;
  };

  ConsoleReporter.prototype.finishTest = function(test, err) {
    var check, message, name;

    name = this.getItemName(test);
    if (!name) {
      return this;
    }
    check = (err ? this.config.fail : this.config.pass);
    message = "" + name + check;
    console.log(message, (typeof process !== "undefined" && process !== null) === false && err ? [err, err.stack] : '');
    return this;
  };

  ConsoleReporter.prototype.exit = function(exitCode) {
    var errorLog, errorLogs, index, name, suite, test, totalErrors, totalFailedTests, totalIncompleteTests, totalPassedTests, totalTests, _i, _len, _ref1, _ref2;

    _ref1 = this.joe.getTotals(), totalTests = _ref1.totalTests, totalPassedTests = _ref1.totalPassedTests, totalFailedTests = _ref1.totalFailedTests, totalIncompleteTests = _ref1.totalIncompleteTests, totalErrors = _ref1.totalErrors;
    if (exitCode) {
      errorLogs = this.joe.getErrorLogs();
      console.log("\n" + this.config.summaryFail, totalPassedTests, totalTests, totalFailedTests, totalIncompleteTests, totalErrors);
      for (index = _i = 0, _len = errorLogs.length; _i < _len; index = ++_i) {
        errorLog = errorLogs[index];
        suite = errorLog.suite, test = errorLog.test, name = errorLog.name, err = errorLog.err;
        name || (name = this.getItemName(test || suite));
        console.log("\n" + this.config.failHeading, index + 1);
        console.log(name);
        console.log(((_ref2 = err.stack) != null ? _ref2.toString() : void 0) || err);
      }
      console.log('');
    } else {
      console.log("\n" + this.config.summaryPass, totalPassedTests, totalTests);
    }
    return this;
  };

  return ConsoleReporter;

})();

module.exports = ConsoleReporter;
